sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/DataSource.scala

```diff
@@ -468,7 +489,13 @@ case class DataSource(
       throw new AnalysisException("Cannot save interval data type into external storage.")
     }
 
-    providingClass.newInstance() match {
+    val resolvedRelation = providingClass.newInstance() match {
+      case relationToCheck: DataSourceValidator =>
```

Comment:
can we do the check at https://github.com/apache/spark/pull/17758/files#diff-7a6cb188d2ae31eb3347b5629a679cecR325 ?

Basically
```
val r = dataSource.createRelation(sparkSession.sqlContext, caseInsensitiveOptions)
SchemaUtils.checkSchemaColumnNameDuplication(r.schema)
r
```